reactive/docker.py

```diff
@@ -257,55 +254,88 @@ def install_from_upstream_apt():
-        deb.append('deb {0}/{1}/ubuntu{2}/{3} /'.format(nvidurl, i, rel, arch))
+        deb.append('deb {0}/{1}/ubuntu{2}/{3} /'.format(nvidurl,
+                                                        i, rel, architecture))
+
+    write_docker_sources(deb)
+
+    install_cuda_drivers_repo(architecture, rel, ubuntu)
+
+    apt_update(fatal=True)
+    # actually install the required packages docker-ce nvidia-docker2
+
+    docker_ce = hookenv.config('docker-ce-package')
+    nvidia_docker2 = hookenv.config('nvidia-docker-package')
+    nv_container_runtime = hookenv.config('nvidia-container-runtime-package')
+    apt_install(['cuda-drivers', docker_ce, nvidia_docker2,
+                 nv_container_runtime], fatal=True)
+
+
+def install_cuda_drivers_repo(architecture, rel, ubuntu):
+    ''' Install cuda drivers this is xenial only.
+     We want to install cuda-drivers only this means that the
+     cuda version plays no role. Any repo will do. '''
+    key_file = '7fa2af80.pub'
+    # distribution should be something like ubuntu1604
+    distribution = '{}{}'.format(ubuntu, str(rel).replace('.', ''))
+    repo_path = 'developer.download.nvidia.com/compute/' \
+                'cuda/repos/{}/x86_64'.format(distribution)
+    cmd = 'apt-key adv --fetch-keys http://{}/{}'.format(repo_path, key_file)
+    check_call(split(cmd))
+    cuda_repo_version = config('cuda_repo')
+    cuda_repo_pkg = 'cuda-repo-{}_{}_{}.deb'.format(distribution,
+                                                    cuda_repo_version,
+                                                    architecture)
+    repo_url = 'https://{}/{}'.format(repo_path, cuda_repo_pkg)
+    r = requests.get(repo_url)
+    r.raise_for_status()
+    with open(cuda_repo_pkg, "wb") as repo_file:
+        for chunk in r.iter_content(chunk_size=1024):
+            repo_file.write(chunk)
+    r.close()
+    cmd = 'dpkg -i  {}'.format(cuda_repo_pkg)
+    check_call(split(cmd))
+
 
+def write_docker_sources(deb):
+    '''Write docker.list under etc/apt/sources.list.d'''
     # mkdir -p /etc/apt/sources.list.d
     if not os.path.isdir('/etc/apt/sources.list.d'):
         os.makedirs('/etc/apt/sources.list.d')
     # Write the docker source file to the apt sources.list.d directory.
     with(open('/etc/apt/sources.list.d/docker.list', 'w+')) as stream:
         stream.write("\n".join(deb))
 
-    apt_update(fatal=True)
-
-    # actually install the required packages docker-ce nvidia-docker2
-    apt_install(['docker-ce', 'nvidia-docker2'], fatal=True)
-
-    for m in ['nvidia_drm', 'nvidia_uvm', 'nvidia_modeset', 'nvidia']:
-        try:
-            hookenv.log('rmmod {0}'.format(m))
-            check_call(['rmmod', m])
-        except CalledProcessError:
-            hookenv.log('not unloading kmod {0}'.format(m))
 
-    try:
-        check_call(['modprobe', 'nvidia'])
-    except CalledProcessError:
-        hookenv.log('not reloading nvidia kmod')
+def add_apt_key(key):
```

Comment:
Similar comment as above -- if you go with the `add-source` charmhelper, your key can be included without having to handle `apt-key adv` yourself.